metricbeat/module/system/cpu/metrics/metrics.go

```diff
@@ -0,0 +1,249 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package metrics
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/v7/libbeat/common"
+)
+
+// CPU manages the CPU metrics from /proc/stat
+// If a given metric isn't available on a given platform,
+// The value will be null. All methods that use these fields
+// should assume that any value can be null.
+// The values are in "ticks", which translates to milliseconds of CPU time
+type CPU struct {
+	user    *uint64 `metric:"user"`
+	sys     *uint64 `metric:"system"`
+	idle    *uint64 `metric:"idle"`
+	nice    *uint64 `metric:"nice"`    // Linux, Darwin, BSD
+	irq     *uint64 `metric:"irq"`     // Linux and openbsd
+	wait    *uint64 `metric:"iowait"`  // Linux and AIX
+	softIrq *uint64 `metric:"softirq"` // Linux only
+	stolen  *uint64 `metric:"steal"`   // Linux only
+}
+
+// CPUMetrics carries global and per-core CPU metrics
+type CPUMetrics struct {
+	totals CPU
+	// list carries the same data, broken down by CPU
+	list []CPU
+}
+
+// Total returns the total CPU time in ticks as scraped by the API
+func (cpu CPU) Total() uint64 {
+
+	var total uint64
+	fn := func(field reflect.Value, _ int, _ string) {
+		total = total + field.Uint()
```

Comment:
`field.Uint` can panic if the type can't be converted.

Reply:
Is there any circumstances where that'll happen? The fields are entirely uint64, and the previous lines check to see if the value is `nil`.

Reply:
If someone changes the types, or introduces new ones, yes it will panic. By moving to reflection you give up type safety on purpose. It is not much different from using a type case on `interface{}` without checking if the cast can succeed.

Reply:
> In this case, is total really correct by blindly summing up all fields?

This is logic carried over from `gosigar`, other components assume `Total()` is more a measure of how much time has passed relative to the CPU, the actual total time reported to metricbeat subtracts idle time from that total.

> If someone changes the types, or introduces new ones, yes it will panic.

Yah, this is currently built off the assumption that everything here is a `uint64`, which is a fairly safe assumption for CPU data so far across the OSes.